tests/data/model-validation.ts

```diff
@@ -135,4 +135,647 @@ type org
+        source: "linter",
+        startColumn: 12,
+        startLineNumber: 7,
+      },
+      {
+        endColumn: 18,
+        endLineNumber: 8,
+        message: "`viewer` is an impossible relation (no entrypoint).",
+        relatedInformation: {
+          relation: "viewer",
+          type: "relation-no-entry-point",
+        },
+        severity: 8,
+        source: "linter",
+        startColumn: 12,
+        startLineNumber: 8,
+      },
+    ],
+  },
+  {
+    // note that for now we have to add sub-document to trick recognition as version 1.1
+    name: "cyclic loop",
+    friendly: `type document
+  relations
+    define reader as writer
+    define writer as reader
+type subdocument
+  relations
+    define parent: [document] as self
+`,
+    expectedError: [
+      {
+        endColumn: 18,
+        endLineNumber: 3,
+        message: "`reader` is an impossible relation (no entrypoint).",
+        relatedInformation: {
+          relation: "reader",
+          type: "relation-no-entry-point",
+        },
+        severity: 8,
+        source: "linter",
+        startColumn: 12,
+        startLineNumber: 3,
+      },
+      {
+        endColumn: 18,
+        endLineNumber: 4,
+        message: "`writer` is an impossible relation (no entrypoint).",
+        relatedInformation: {
+          relation: "writer",
+          type: "relation-no-entry-point",
+        },
+        severity: 8,
+        source: "linter",
+        startColumn: 12,
+        startLineNumber: 4,
+      },
+    ],
+  },
+  {
+    name: "parent relation used inside contains a write",
+    friendly: `type user
+type folder
+  relations
+    define parent: [folder] as self or parent from parent
+    define viewer: [user] as self or viewer from parent
+`,
+    expectedError: [
+      {
+        endColumn: 58,
+        endLineNumber: 4,
+        message: "`parent` relation used inside from allows only direct relation.",
+        relatedInformation: {
+          relation: "parent",
+          type: "tupleuset-not-direct",
+        },
+        severity: 8,
+        source: "linter",
+        startColumn: 52,
+        startLineNumber: 4,
+      },
+      {
+        endColumn: 56,
+        endLineNumber: 5,
+        message: "`parent` relation used inside from allows only direct relation.",
+        relatedInformation: {
+          relation: "parent",
+          type: "tupleuset-not-direct",
+        },
+        severity: 8,
+        source: "linter",
+        startColumn: 50,
+        startLineNumber: 5,
+      },
+    ],
+  },
+  {
+    name: "parent relation used inside viewer contains a write",
+    friendly: `type user
+type folder
+  relations
+    define root: [folder] as self
+    define parent: [folder] as self or root
+    define viewer: [user] as self or viewer from parent
+`,
+    expectedError: [
+      {
+        endColumn: 56,
+        endLineNumber: 6,
+        message: "`parent` relation used inside from allows only direct relation.",
```

Comment:
```suggestion
        message: "`parent` relation must be defined `as self`.",
```

Reply:
to be defer in a later PR when there are consensus in the wordings